tests/optimize/test_spsa.py

```diff
@@ -0,0 +1,337 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for the SPSA optimizer"""
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+univariate = [(np.sin), (lambda x: np.exp(x / 10.0)), (lambda x: x**2)]
+
+multivariate = [
+    (lambda x: np.sin(x[0]) + np.cos(x[1])),
+    (lambda x: np.exp(x[0] / 3) * np.tanh(x[1])),
+    (lambda x: np.sum([x_**2 for x_ in x])),
+]
+
+
+class TestSPSAOptimizer:
+    """Test the SPSA optimizer"""
+
+    @pytest.mark.parametrize("args", [0, -3, 42])
+    @pytest.mark.parametrize("f", univariate)
+    def test_apply_grad(self, args, f):
+        """Test that a gradient step can be applied correctly with a univariate
+        function."""
+        spsa_opt = qml.SPSAOptimizer(maxiter=10)
+        args = np.array([args], requires_grad=True)
+
+        alpha = 0.602
+        gamma = 0.101
+        c = 0.2
+        A = 20.0
+        a = 0.05 * (A + 1) ** alpha
+        k = 1
+        ck = c / (k + 1.0) ** gamma
+        ak = a / (A + k + 1.0) ** alpha
+        tol = np.maximum(np.abs(f(args - ck)), np.abs(f(args + ck)))
+
+        y = f(args)
+        grad = (y) / (2 * ck)
+
+        res = spsa_opt.apply_grad(grad, args)
+        expected = args - ak * grad
+        assert np.allclose(res, expected, atol=tol)
+
+    @pytest.mark.parametrize("f", multivariate)
+    def test_apply_grad_multivar(self, f):
+        """
+        Test that a gradient step can be applied correctly to a multivariate
+        function.
+        """
+        alpha = 0.602
+        gamma = 0.101
+        c = 0.2
+        A = 20.0
+        a = 0.05 * (A + 1) ** alpha
+        k = 1
+        ck = c / (k + 1.0) ** gamma
+        ak = a / (A + k + 1.0) ** alpha
+        spsa_opt = qml.SPSAOptimizer(maxiter=10)
+
+        x_vals = np.linspace(-10, 10, 16, endpoint=False)
+
+        for jdx in range(len(x_vals[:-1])):
+            x_vec = x_vals[jdx : jdx + 2]
+            y = f(x_vec)
+            grad = (y) / (2 * ck * np.ones((2)))
+            x_new = spsa_opt.apply_grad(grad, x_vec)
+            x_al = x_vec - ak * grad
+            tol = np.maximum(np.abs(f(x_vec - ck)), np.abs(f(x_vec + ck)))
+            assert np.allclose(x_new, x_al, atol=tol)
+
+    @pytest.mark.parametrize("args", [0, -3, 42])
+    @pytest.mark.parametrize("f", univariate)
+    def test_step_and_cost_supplied_cost(self, args, f):
+        """Test that returned cost is correct"""
+        spsa_opt = qml.SPSAOptimizer(maxiter=10)
+        args = np.array(args, requires_grad=True)
+
+        _, res = spsa_opt.step_and_cost(f, args)
+        expected = f(args)
+        assert np.all(res == expected)
+
+    def test_step_and_cost_supplied_cost2(self):
```

Comment:
Worth updating the test names such that it describes the test case more precisely. That way no numbers would have to be used.